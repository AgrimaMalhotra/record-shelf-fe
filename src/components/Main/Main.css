.main-body {
    margin: 0 15vw 0 15vw;
    background-color: #f5f5f5;
    height: 100%;
    padding: 7vh 10vw 0 10vw;
}

.main-body-header {
    display: flex;
    justify-content: space-between;
    width: 100%;
    margin-bottom: 25px;
    margin-top: 25px;
}

.main-body-header>p {
    font-size: 6vh;
    font-family: Arial Black;
}

.main-body-header>img {
    margin-top: 25px;
    height: 4vh;
    cursor: pointer;
}

.genre {
    position: relative;
    height: 170px;
}

.genre>p {
    font-size: 30px;
    font-family: Arial Black;
    background-color: #4C63AA;
    color: white;
    height: 50px;
    width: 250px;
    border-radius: 60px;
    text-transform: lowercase;
    text-align: center;
    position: absolute;
    top: 50%;
    left: 220px;
    transform: translate(-50%, -50%);
}

@media only screen and (max-width: 700px) {
    .main-body {
        margin: 0;
    }
}

/* 
@media only screen and (min-width: 480px) and (min-height) {
    .main-body-header {
        margin-top: 25px;
    }
} */