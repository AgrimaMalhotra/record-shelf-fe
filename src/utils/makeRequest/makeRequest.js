import axios from 'axios';
import { BACKEND_URL } from '../../constants/apiEndPoints';
import { ERROR_ROUTE } from '../../constants/routes';

const makeRequest = async (apiEndPoint, dynamicConfig, navigate) => {
  try {
    const requestDetails = {
      baseURL: BACKEND_URL,
      url: apiEndPoint.url,
      method: apiEndPoint.method,
      headers: {
        authorization: 'Bearer QWlzaHdhcnlhIE4=',
      },
      ...dynamicConfig,
    };
    const { data } = await axios(requestDetails);
    return data;
  } catch (error) {
    if (navigate) {
      const errorCode = error.response?.status;
      if (errorCode) {
        navigate(`${ERROR_ROUTE}/${errorCode}`);
      } else {
        navigate({ ERROR_ROUTE });
      }
    }
  }
};
export default makeRequest;
