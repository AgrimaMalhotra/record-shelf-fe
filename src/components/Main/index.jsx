import React, { useState } from 'react';
import CardGrid from '../CardGrid';
import './Main.css';
import genreIcon from '../../assets/icon-genre.svg';
import gridIcon from '../../assets/icon-grid.svg';
import GENRE_IMAGE_MAPPING from '../../constants/imageMapping';
import propTypes from 'prop-types';

const Main = ({ allSongsData }) => {
  const [clickedGenreIcon, setClickedGenreIcon] = useState(false);

  const handleClick = () => {
    setClickedGenreIcon(!clickedGenreIcon);
  };

  const getGenreClassification = () => {
    return allSongsData.reduce((acc, song) => {
      if (acc[song.genre['name']]) {
        acc[song.genre['name']].push(song);
      } else {
        acc[song.genre['name']] = [song];
      }
      return acc;
    }, {});
  };
  const genreClassifiedData = getGenreClassification();

  return (
    <div className="main-body">
      {clickedGenreIcon ? (
        <>
          <div className="main-body-header">
            <p>genres</p>
            <img src={gridIcon} alt="Grid Icon" onClick={handleClick} />
          </div>
          <div className="genre-header">
            {Object.keys(genreClassifiedData).map((key) => (
              <div key={key}>
                <div className="genre">
                  <img src={GENRE_IMAGE_MAPPING[key]} alt="Genre" />
                  <p className="genre-name">{key}</p>
                </div>
                <CardGrid
                  className="grid"
                  allSongsData={genreClassifiedData[key]}
                />
              </div>
            ))}
          </div>
        </>
      ) : (
        <>
          <div className="main-body-header" data-testid="test-all-songs-page">
            <p>all songs</p>
            <img
              src={genreIcon}
              alt="Grid Icon"
              onClick={handleClick}
              data-testid="test-genre-icon-button"
            />
          </div>
          <CardGrid className="grid" allSongsData={allSongsData} />
        </>
      )}
    </div>
  );
};

Main.propTypes = {
  allSongsData: propTypes.array.isRequired,
};
export default Main;
