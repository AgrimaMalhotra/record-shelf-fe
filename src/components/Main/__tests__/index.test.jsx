import {render, screen, waitFor} from '@testing-library/react';
import * as React from 'react';
import { useNavigate } from 'react-router-dom';
import Main from '..';
import makeRequest from '../../../utils/makeRequest/makeRequest';
import { mockAllSongsData, mockJustSongDetail } from '../../../mocks/allSongsData';
jest.mock('react-router-dom',()=>({useNavigate:()=>jest.fn()})
);
jest.mock('../../../utils/makeRequest/makeRequest');
describe('Main',()=>{
  it('should render correctly',()=>{
    const {container}= render(<Main/>);
    expect(container).toMatchSnapshot();
  });
  it('should show loading when data is still being fetched',()=>{
makeRequest.mockResolvedValueOnce(mockJustSongDetail);
const fetchLikesData = jest.fn();
fetchLikesData.mockResolvedValueOnce(mockAllSongsData)
    render(<Main/>);
await waitFor(()=>{
  expect(screen.getByText('all songs')).
})
  })
})