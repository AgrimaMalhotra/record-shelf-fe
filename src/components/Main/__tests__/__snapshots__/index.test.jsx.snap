// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render correctly 1`] = `
<div>
  <div
    class="main-body"
  >
    <div
      class="main-body-header"
    >
      <p>
        all songs
      </p>
      <img
        alt="Grid Icon"
        src="icon-genre.svg"
      />
    </div>
    <div
      class="card-grid"
    />
  </div>
</div>
`;
